ipi/engine/forcefields.py

```diff
@@ -1371,6 +1374,162 @@ def poll(self):
                     r["status"] = "Done"
 
 
+class FFRotations(FFSocket):
+    """Forcefield to manipulate models that are not exactly rotationally equivariant.
+    Can be used to evaluate a different random rotation at each evaluation, or to average
+    over a regular grid of Euler angles"""
+
+    def __init__(
+        self,
+        latency=1.0,
+        offset=0.0,
+        name="",
+        pars=None,
+        dopbc=True,
+        active=np.array([-1]),
+        threaded=True,
+        interface=None,
+        random=False,
+        inversion=False,
+        grid=1,
+    ):
+        super(FFRotations, self).__init__(
+            latency,
+            offset,
+            name,
+            pars,
+            dopbc,
+            active,
+            threaded,
+            interface,
+        )
+
+        # TODO: initialize and save (probably better to use different PRNG than the one from the simulation)
+        self.prng = Random()
+        self.random = random
+        self.inversion = inversion
+        self.grid = grid
+
+        self._rotations = get_rotation_quadrature(self.grid)
+
+    def queue(self, atoms, cell, reqid=-1):
+        # launches requests for all of the rotations FF objects
+        ffh = []
+        rots = []
+        if self.random:
+            R_random = random_rotation(self.prng, improper=True)
+        else:
+            R_random = np.eye(3)
+
+        for R, w in self._rotations:
+            R = R @ R_random
+            rot_atoms = atoms.clone()
+            # NB we need generic cell orientation
+            rot_cell = GenericCell(R @ dstrip(cell.h).copy())
+            rot_atoms.q[:] = (dstrip(rot_atoms.q).reshape(-1, 3) @ R.T).flatten()
```

Comment:
This seems like it does the correct thing of doing `R @ pos` where `pos` is column-wise the positions. Just checking.

Reply:
yes, it does the right thing with q being stored with x,y,z in rows